File: 2026-02/src/day_25/boj_16934/Main.cpp
#include <bits/stdc++.h>
using namespace std;

const int MX = 1 + 100000 * 10;
const int ROOT = 0;
int unused = ROOT + 1;
int nxt[MX][26];
unordered_map<string, int> mp;  // key를 닉네임으로 가입한 유저의 수(value)

void insert(string& s) {
    int cur = ROOT;
    for (char c : s) {
        int idx = c - 'a';
        if (nxt[cur][idx] == 0) nxt[cur][idx] = unused++;
        cur = nxt[cur][idx];
    }
}

string find(string& s) {
    string prefix = "";
    int cur = ROOT;
    for (char c : s) {
        prefix += c;
        int idx = c - 'a';
        if (nxt[cur][idx] == 0) return prefix;
        cur = nxt[cur][idx];
    }

    if (mp[s] > 0) prefix += to_string(mp[s] + 1);
    return prefix;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n;
    cin >> n;

    while (n--) {
        string s;
        cin >> s;

        cout << find(s) << '\n';
        insert(s);
        mp[s]++;
    }
}

File: 2026-02/src/day_25/boj_16934/Main.java
package day_25.boj_16934;

import java.io.*;
import java.util.*;

public class Main {

    static final int MX = 1 + 100000 * 10;
    static final int ROOT = 0;
    static int unused = ROOT + 1;
    static int[][] nxt = new int[MX][26];
    static Map<String, Integer> map = new HashMap<>();  // key를 닉네임으로 가입한 유저의 수(value)

    static void insert(String s) {
        int cur = ROOT;
        for (char c : s.toCharArray()) {
            int idx = c - 'a';
            if (nxt[cur][idx] == 0) nxt[cur][idx] = unused++;
            cur = nxt[cur][idx];
        }
    }

    static String find(String s) {
        StringBuilder prefix = new StringBuilder();
        int cur = ROOT;
        for (char c : s.toCharArray()) {
            prefix.append(c);
            int idx = c - 'a';
            if (nxt[cur][idx] == 0) return prefix.toString();
            cur = nxt[cur][idx];
        }

        if (map.containsKey(s)) prefix.append(map.get(s) + 1);
        return prefix.toString();
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int n = Integer.parseInt(br.readLine());
        while (n-- > 0) {
            String s = br.readLine();

            sb.append(find(s)).append("\n");
            insert(s);
            map.put(s, map.getOrDefault(s, 0) + 1);
        }

        System.out.println(sb);
    }
}
